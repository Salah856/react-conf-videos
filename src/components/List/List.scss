.root {
  margin: 0;
  padding: 0 0 2rem 0;
  list-style: none;
}
