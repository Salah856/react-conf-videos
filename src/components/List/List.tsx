import * as React from 'react'
import VirtualList from 'react-virtual-list'
import { mapProps } from 'recompose'
import { compose as rcompose, flatten, map, pathOr, toPairs } from 'ramda'

import { Video } from 'components'
import { Conference, IndexedConferences } from '../../domain'

import styles from './List.scss'

type MappedVideo = {
  key: string,
  videoId: string,
  conferenceId: string
}

type Props = { conferences: {[idx: string]: Conference} }
type MapProps = { videos: MappedVideo[] }

const mapConferenceIdOntoVideos = ([conferenceId, conference]: [string, Conference]) =>
  map(
    (video) => ({ key: video, videoId: video, conferenceId }),
    pathOr([], ['videos'], conference)
  )

const mapConferenceVideos = rcompose<IndexedConferences, any, any[], MappedVideo[]>(
  flatten,
  map(mapConferenceIdOntoVideos),
  toPairs
)

const VirtualisedList = ({ virtual }: {virtual: any}): React.ReactElement<{}> => (
  <div style={virtual.style}>
    {virtual.items && virtual.items.map((item: any) => (
      <Video {...item} />
    ))}
  </div>
)

const MyVirtualList = VirtualList({
  firstItemIndex: 0,
  lastItemIndex: 20,
  initialState: {
    firstItemIndex: 0,
    lastItemIndex: 20,
  }
})(VirtualisedList);

export const ListInner: React.SFC<MapProps> = ({ videos }) => {
  return (
    <section className={styles.root}>
      {videos.length > 0 && <MyVirtualList
        items={videos}
        itemHeight={60}
        itemBuffer={20}
      /> }
    </section>
  )
}

const List = mapProps<MapProps, Props>(({conferences}) =>
  ({ videos: mapConferenceVideos(conferences)})
)(ListInner)

export default List
