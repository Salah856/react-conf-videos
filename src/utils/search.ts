import { compose, toLower, trim } from 'ramda'
import { remove as removeDiacritics } from 'diacritics'

export const cleanQuery = compose(
  removeDiacritics,
  toLower,
  trim
)

export const isFilterEmpty = (filterValue: string = '') => filterValue.trim() === ''
