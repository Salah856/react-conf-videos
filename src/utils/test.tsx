import { mount, shallow } from 'enzyme';
import * as React from 'react';
import { ReactElement } from 'react';
import { MemoryRouter } from 'react-router';
import { StateObservable } from 'redux-observable';
import { createMockStore } from 'redux-test-utils';
import { Subject } from 'rxjs';
import * as Immutable from 'seamless-immutable';

import { ApplicationState } from 'redux/modules';
import {
  Conference,
  IndexedConferences,
  IndexedPresenters,
  IndexedVideos,
  Presenter,
  Video
} from '../domain';

// cast to ApplicationState so TS doesn't complain when sending in partials to Epics for testing
export const mockStore = (state = {}) =>
  new StateObservable<ApplicationState>(
    new Subject(),
    state as ApplicationState
  );

const isStore = (store: any) => store.hasOwnProperty('isActionTypeDispatched');

// shallow mount component with a store in context
export const shallowWithStore = (storeP: any, component: ReactElement<{}>) => {
  const store = isStore(storeP) ? storeP : createMockStore(storeP);
  return shallow<any, any>(component, { context: { store } });
};

export const mountWithStore = (storeP: any, component: ReactElement<{}>) => {
  const store = isStore(storeP) ? storeP : createMockStore(storeP);
  return mount<any, any>(component, { context: { store } });
};

export const wrapWithMemoryRouter = (Component: React.ReactElement<any>) => (
  // key within location prop keeps changing, so manually set
  // so snapshorts are consistent
  <MemoryRouter initialEntries={[{ key: 'xxxx' }]}>{Component}</MemoryRouter>
);

/** Mock data */
export const mockVideo = (title = 'test title'): Video => ({
  link: 'a link',
  split: '',
  title: title || 'test title',
  length: '12:34',
  lightning: false,
  presenter: 'siomn carter',
  embeddableLink: 'a link'
});

export const mockVideos = (): IndexedVideos => ({
  aaa: mockVideo('aaa'),
  bbb: mockVideo('bbb'),
  ccc: mockVideo('ccc')
});

export const mockPresenter = (name = 'simon carter'): Presenter => ({
  name
});

export const mockPresenters = (): IndexedPresenters => ({
  xxx: mockPresenter(),
  zzz: mockPresenter('zzz')
});

export const mockConference = (): Conference => ({
  date: 'XX/YY/ZZZ',
  title: 'fake conf title',
  website: 'fake url',
  playlist: 'day 1',
  videos: ['aaa', 'bbb', 'ccc']
});

export const mockIndexedConferences = (): IndexedConferences => ({
  ['XXXX']: mockConference()
});

export const mockRouterState = (
  pathname = '/conference/8424e37df85b9eccbe48e9a55d93845e/react-conf-2018'
) => ({
  router: {
    location: {
      pathname,
      search: '',
      hash: ''
    }
  }
});

export const mockConferencePageSlice = () => ({
  conferencePage: {
    selectedConferenceId: 'XXXX',
    conference: mockConference()
  }
});

export const mockDataState = () =>
  Immutable({
    data: {
      conferences: mockIndexedConferences(),
      presenters: mockPresenters(),
      videos: mockVideos()
    }
  });
