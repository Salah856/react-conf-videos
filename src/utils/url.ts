import { compose, replace, toLower } from 'ramda';

// replaces white space with - and lower case
export const sluggifyUrl = compose(
  replace(/\s/g, '-'),
  toLower
);
