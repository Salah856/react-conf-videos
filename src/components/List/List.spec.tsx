import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import * as React from 'react';

import { mockConference } from 'utils/test';

import { List } from './List';

describe('List', () => {
  it('should render', () => {
    // arrange
    const conf = mockConference();
    const props = {
      conferences: { yyy: conf }
    };

    // act
    const wrapper = shallow(<List {...props} />);

    // assert
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
