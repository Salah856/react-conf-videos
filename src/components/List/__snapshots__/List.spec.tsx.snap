// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render 1`] = `
<section
  className="root"
>
  <vlist
    itemBuffer={20}
    itemHeight={60}
    items={
      Array [
        Object {
          "conferenceId": "yyy",
          "key": "aaa",
          "videoId": "aaa",
        },
        Object {
          "conferenceId": "yyy",
          "key": "bbb",
          "videoId": "bbb",
        },
        Object {
          "conferenceId": "yyy",
          "key": "ccc",
          "videoId": "ccc",
        },
      ]
    }
  />
</section>
`;
